lighthouse-core/report/v2/renderer/report-renderer.js

```diff
@@ -22,30 +22,48 @@
  * Dummy text for ensuring report robustness: </script> pre$`post %%LIGHTHOUSE_JSON%%
  */
 
-/* globals self */
+/* globals self, formatDateTime, formatNumber, calculateRating */
 
 class ReportRenderer {
   /**
    * @param {!DOM} dom
    * @param {!CategoryRenderer} categoryRenderer
+   * @param {ReportUIFeatures=} uiFeatures
    */
-  constructor(dom, categoryRenderer) {
+  constructor(dom, categoryRenderer, uiFeatures = null) {
     /** @private {!DOM} */
     this._dom = dom;
     /** @private {!CategoryRenderer} */
     this._categoryRenderer = categoryRenderer;
+    /** @private {!Document|!Element} */
+    this._templateContext = this._dom.document();
+    /** @private {ReportUIFeatures} */
+    this._uiFeatures = uiFeatures;
   }
 
   /**
    * @param {!ReportRenderer.ReportJSON} report
+   * @param {!Element} container Parent element to render the report into.
    * @return {!Element}
    */
-  renderReport(report) {
+  renderReport(report, container) {
+    container.textContent = ''; // Remove previous report.
+
+    let element;
     try {
-      return this._renderReport(report);
+      element = container.appendChild(this._renderReport(report));
+
+      // Hook in JS features and page-level event listeners after the report
+      // is in the document.
+      if (this._uiFeatures) {
+        this._uiFeatures.initFeatures(report, container);
+      }
     } catch (/** @type {!Error} */ e) {
-      return this._renderException(e);
+      container.textContent = '';
+      element = container.appendChild(this._renderException(e));
     }
+
+    return /** @type {!Element} **/ (element);
```

Comment:
can also split into `element = this._renderReport(report); container.appendChild(element);` lines if you don't want to cast